updated to build framework-arduinostm32mxchip v2.0.0

fmuntean opened the pull request:
updated and submitted the framework-arduinostm32mxchip package for approval to PlatformIO.
The mxchip team changed the framework drastically and the existing build script does not work.
I have included a build script in the framework package itself so we have to change this script to check if a build script exists in the package itself then use it. For backwards compatibility I left the existing script to run for older frameworks.

fmuntean commented:
Fixes: #273

valeros commented:
Many thanks for your PR. I've updated the build script according to the latest changes in the framework. Please retest with the upstream platform:
```
[env:mxchip_az3166]
platform = https://github.com/platformio/platform-ststm32.git
framework = arduino
board = mxchip_az3166
```

fmuntean commented:
> Many thanks for your PR. I've updated the build script according to the latest changes in the framework. Please retest with the upstream platform:
> 
> ```
> [env:mxchip_az3166]
> platform = https://github.com/platformio/platform-ststm32.git
> framework = arduino
> board = mxchip_az3166
> ```

it still loads the old package version:
Processing test-mxchip_az3166 (platform: https://github.com/platformio/platform-ststm32.git; framework: arduino; board: mxchip_az3166)
--------------------------------------------------------------------------------------------------------------------------------------------------------Tool Manager: Installing platformio/framework-arduinostm32mxchip @ ~1.3.0
Unpacking  [####################################]  100%
Tool Manager: framework-arduinostm32mxchip @ 1.3.2 has been installed!

fmuntean commented:
using the following environment config:
[env:test-mxchip_az3166]
platform = https://github.com/platformio/platform-ststm32.git
platform_packages = framework-arduinostm32mxchip@2.0.0
framework = arduino
board = mxchip_az3166

the downlaoded package is missing the platformio-build.py file.
It seems to download the wrong package as the package: 9c55dd7-framework-arduinostm32mxchip-2.0.0.tar.gz has the file inside it.

valeros commented:
> the downloaded package is missing the platformio-build.py file.

Looks like you've mixed your packages with the official shipped with the `ststm32` platform.
The original one shouldn't contain any build scripts as they are present in the platform.

> How do I use the new framework-arduinostm32mxchip@2.0.0 ?

Probably, you've previously installed the upstream version of the platform and it's outdated. Make sure your platforms are up to date via running `pio update` in the IDE terminal (keep in mind that PlatformIO may override your local changes in the packages)

fmuntean commented:
I don't understand on the url: https://bintray.com/platformio/tool-packages/framework-arduinostm32mxchip#files
now shows only one file for the 2.0.0 and is the wrong one.
The correct one which was removed was: 9c55dd7-framework-arduinostm32mxchip-2.0.0.tar.gz

Please advise

fmuntean commented:
> Looks like you've mixed your packages with the official shipped with the `ststm32` platform.
> The original one shouldn't contain any build scripts as they are present in the platform.

There was not updated version for mxchip 2.0.0 and this is why I created the pull request and published the framework-arduinostm32mxchip@2.0.0 package.
Somehow the one I published (9c55dd7-framework-arduinostm32mxchip-2.0.0.tar.gz) got overwritten by another one instead.
Please advise on how we fix this.

valeros commented:
There is no need to maintain two packages, the last one `8b53f03-framework-arduinostm32mxchip-2.0.0.tar.gz` is the package used in this platform.

fmuntean commented:
> There is no need to maintain two packages, the last one `8b53f03-framework-arduinostm32mxchip-2.0.0.tar.gz` is the package used in this platform.

I don't think is the one I published. How this package was build? This package is not building due to the major changes inside the MXChip repository when upgrading from 1.3.2 to 2.0.0 and this is why I included the platformio-build.py inside my package.
Se repository to use to create the package needs to be: https://github.com/fmuntean/devkit-sdk/tree/mfd/AZ3166/src
This is a fork of the official repository and I have a pull request with Microsoft on this to push it upstream.

valeros commented:
Your package is no longer available, Also, you don't need this fork anymore since we've updated build script directly in the platform. You need to use the upstream version of the platform.

fmuntean commented:
> Your package is no longer available. Also, you don't need this fork anymore since we've updated build script directly in the platform. You need to use the upstream version of the platform.

I don't understand. 
Can you explain the process?
Currently I see none of my changes make it trough and the new package for framework-arduinostm32mxchip is not the one I pushed with the correct build script inside.

So as we stand I can't build using the new framework.
----

I want to reiterate here what I have done:
1. in order to use the new framework the system requires a new build python file due to major changes
2. in order to keep it decoupled and backwards compatible I changed the mxchip.py under platformio-stm32 to check for a python file (platformio-build.py) under the framework-arduinostm32mxchip package and execute that if exists otherwise use the old process so people can still use the old framework version.
3. published the correct package with the platformio-build.py to the platformio repository.
with all these in place we should be able to build using either v1.3.2 or v2.0.0
----


Please advise.

fmuntean commented:
Is there a faster way, than just waiting for each other comments, so we can get together to sort this issue out and resolve it ?

valeros commented:
Again, here are the main points:
- Your package (with bundled `platformio-build.py`) is no longer available, it has been replaced by the original package v2.0.0 from https://github.com/microsoft/devkit-sdk and build script has been updated accordingly here https://github.com/platformio/platform-ststm32/commit/f56c535d6e691013432a3201dcb00850e3e77a75
- Since the latest changes hasn't been released yet, you need install the upstream version of the platform from dev branch

I've a suspicion that you have multiple `ststm32` platform installed in your system. I propose the following steps:
- Manually delete all installed `ststm32` platforms in PlatformIO Home directory (`C:\Users\YOUR_USER\.platformio\platforms` on Windows or `/home/your_user/.platformio/platforms` on Linux)
- Run `pio platform install https://github.com/platformio/platform-ststm32.git` from the IDE terminal

Now the latest platform install you can set environments as usual:
```
[env:test-mxchip_az3166]
platform = ststm32
framework = arduino
board = mxchip_az3166
```

fmuntean commented:
> Again, here are the main points:
> 
> * Your package (with bundled `platformio-build.py`) is no longer available, it has been replaced by the original package v2.0.0 from https://github.com/microsoft/devkit-sdk and build script has been updated accordingly here [f56c535](https://github.com/platformio/platform-ststm32/commit/f56c535d6e691013432a3201dcb00850e3e77a75)
> * Since the latest changes hasn't been released yet, you need install the upstream version of the platform from dev branch
> 
> I've a suspicion that you have multiple `ststm32` platform installed in your system. I propose the following steps:
> 
> * Manually delete all installed `ststm32` platforms in PlatformIO Home directory (`C:\Users\YOUR_USER\.platformio\platforms` on Windows or `/home/your_user/.platformio/platforms` on Linux)
> * Run `pio platform install https://github.com/platformio/platform-ststm32.git` from the IDE terminal
> 
> Now the latest platform install you can set environments as usual:
> 
> ```
> [env:test-mxchip_az3166]
> platform = ststm32
> framework = arduino
> board = mxchip_az3166
> ```

I have deleted all then refresh. the problem is still there.
There is no "official 2.0.0" package. I have been asking for that for a while now. So I stepped in and figure out myself how to build it.
The build script was not updated. It sill has the old mxchip.py and my changes are missing.
I am using the dev branch aka "develop" as you name it. 
It downloads the framework you published today but the build is failing.

Please advise what you want me to do on my side.
Once you change the mxchip.py file and delete the wrong framework package from repository I can republish the correct one. 
Or you can republish it using the code from my fork: https://github.com/fmuntean/devkit-sdk/tree/mfd/AZ3166/src

Also please explain how are you building the framework-arduinostm32mxchip as the official repository is missing the required platformio file: package.json 

Let me know what you want to do next.

valeros commented:
> It downloads the framework you published today but the build is failing.

What is the error message? Could you please try to compile a simple example in verbose mode and attach here the output?

> Once you change the mxchip.py file and delete the wrong framework package from repository I can republish the correct one.

We don't use forks for the the official packages used in this platform.

> Also please explain how are you building the framework-arduinostm32mxchip as the official repository is missing the required platformio file: package.json

The `package.json` file is not required to be present in the repository, it's added during publishing it to the CDN on our side.

fmuntean commented:
> > It downloads the framework you published today but the build is failing.
> 
> What is the error message? Could you please try to compile a simple example in verbose mode and attach here the output?
> 
> > Once you change the mxchip.py file and delete the wrong framework package from repository I can republish the correct one.
> 
> We don't use forks for the the official packages used in this platform.
> 
> > Also please explain how are you building the framework-arduinostm32mxchip as the official repository is missing the required platformio file: package.json
> 
> The `package.json` file is not required to be present in the repository, it's added during publishing to the CDN on our side.

From what I can see instead of accepting my changes you updated the build python file yourself.
This results in not making the platform backwards compatible and forcing everybody to use the v2.0.0 so many projects will break instantaneously as the APIs have changed in some parts.
If you still want to go this route here is my platformio-build.py that works. 
[platformio-build.txt](https://github.com/platformio/platform-ststm32/files/5404503/platformio-build.txt)

Please update the repository and let me know when I can test.

valeros commented:
> This results in not making the platform backwards compatible and forcing everybody to use the v2.0.0 so many projects will break instantaneously as the APIs have changed in some parts.

Backwards compatible with what? The support for v2.0.0 will be released together with the major release of the development platform. Anyone who wants to use the previous version of the framework can lock the version of the `ststm32` platform to  `~8.1.0`  directly in the `platformio.ini` file.

> Please update the repository and let me know when I can test.

The repository already contains the updated build script.

---
Attach please here the verbose output with the exact error message you see when trying to compile with the upstream version of the platform.

fmuntean commented:
Processing test-mxchip_az3166 (platform: https://github.com/platformio/platform-ststm32.git; framework: arduino; board: mxchip_az3166)
--------------------------------------------------------------------------------------------------------------------------------------------------------Platform Manager: Installing git+https://github.com/platformio/platform-ststm32.git
git version 2.28.0.windows.1
Cloning into 'C:\Users\florin.muntean\.platformio\.cache\tmp\pkg-installing-9f9mp9i5'...
remote: Enumerating objects: 631, done.
remote: Counting objects: 100% (631/631), done.
remote: Compressing objects: 100% (428/428), done.
remote: Total 631 (delta 405), reused 318 (delta 165), pack-reused 0Receiving objects:  99% (625/631), 1.56 MiB | 1.53 MiB/s
Receiving objects: 100% (631/631), 1.69 MiB | 1.56 MiB/s, done.
Resolving deltas: 100% (405/405), done.
Submodule 'builder/frameworks/libopencm3' (https://github.com/platformio/builder-framework-libopencm3.git) registered for path 'builder/frameworks/libopencm3'
Cloning into 'C:/Users/florin.muntean/.platformio/.cache/tmp/pkg-installing-9f9mp9i5/builder/frameworks/libopencm3'...
remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.        
remote: Compressing objects: 100% (11/11), done.        
remote: Total 41 (delta 18), reused 15 (delta 12), pack-reused 18        
Submodule path 'builder/frameworks/libopencm3': checked out 'c5bf02154273cf2f6b45ddf393480db55b965bb1'
Platform Manager: ststm32 @ 8.1.0+sha.a05ae97 has been installed!
The platform 'https://github.com/platformio/platform-ststm32.git' has been successfully installed!
The rest of the packages will be installed later depending on your build environment.
Tool Manager: Installing platformio/framework-arduinostm32mxchip @ ~2.0.0
Unpacking  [####################################]  100%
Tool Manager: framework-arduinostm32mxchip @ 2.0.0 has been installed!
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/mxchip_az3166.html
PLATFORM: ST STM32 (8.1.0+sha.a05ae97) > Microsoft Azure IoT Development Kit (MXChip AZ3166)
HARDWARE: STM32F412ZGT6 100MHz, 256KB RAM, 1MB Flash
DEBUG: Current (stlink) On-board (stlink) External (blackmagic, jlink)
PACKAGES:
 - framework-arduinostm32mxchip 2.0.0
 - framework-cmsis 2.50501.200527 (5.5.1)
 - toolchain-gccarmnoneeabi 1.60301.0 (6.3.1)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 10 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <Sensor> 0.0.1
|   |-- <WiFi> 1.2.6
|-- <WiFi> 1.2.6
|-- <MQTT Client> 0.0.1
Building in release mode
Compiling .pio\build\test-mxchip_az3166\src\main.cpp.o
Compiling .pio\build\test-mxchip_az3166\src\mqtt.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib2b2\WiFi\AZ3166WiFi.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib2b2\WiFi\AZ3166WiFiClient.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib2b2\WiFi\AZ3166WiFiServer.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib2b2\WiFi\AZ3166WiFiUdp.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib520\Sensors\HTS221Sensor.cpp.o
Compiling .pio\build\test-mxchip_az3166\lib520\Sensors\IoT_DevKit_HW.cpp.o
C:\users\florin.muntean\.platformio\packages\framework-arduinostm32mxchip\libraries\Sensors\src\IoT_DevKit_HW.cpp:2:25: fatal error: AzureIotHub.h: No such file or directory

fmuntean commented:
I think you are missing this in your build script:
for d in ("system", join("cores", env.BoardConfig().get("build.core")), "libraries"):
    for root, _, files in sorted(walk(join(FRAMEWORK_DIR, d))):
        if any(f.endswith(".h") for f in files) or "inc" in root:
            if root not in inc_dirs:
                inc_dirs.append(root)

valeros commented:
The `AzureIotHub.h` file belongs to the `AzureIoT` library. I suspect the problem is that that the default `chain` LDF mode doesn't parse all source files (including `IoT_DevKit_HW.cpp`) and accordingly couldn't find the dependency on the `AzureIoT` library.
Could you try to change the LDF mode in your `platformio.ini`:
```
[env:test-mxchip_az3166]
platform = ststm32
framework = arduino
board = mxchip_az3166

lib_ldf_mode = deep
```

fmuntean commented:
> lib_ldf_mode = deep

using the lib_ldf_mode=deep works however this was not required before nor was required with my build script.
Can you make that the default then for this framework? If not many people will complain that they can't build using the new version.

valeros commented:
> using the lib_ldf_mode=deep works however this was not required before nor was required with my build script.

In your build script you've added library include paths directly to the global build environment. That's not correct, all libraries are dynamically handled by the LDF.

> Can you make that the default then for this framework?

Unfortunately, there is no easy way to do so.

fmuntean commented:
So how we convey to the users that they have to use the lib_ldf_mode = deep ?
Can you update the documentation on platformio board specific page?

valeros commented:
It's possible, before that, could you please attach here a minimal example that fails to compile with `fatal error: AzureIotHub.h: No such file or directory` to reproduce the issue?

fmuntean commented:
[AZ3166_sysinfo.zip](https://github.com/platformio/platform-ststm32/files/5415965/AZ3166_sysinfo.zip)
attached is a project that shows the problem. 
one env config build with deep link one without that it does not.
